### PR TITLE
Init Quantity with numpy array and gpu backend should create gpu storage/array

### PR DESCRIPTION
**Pull request by mcgibbon**
Main changes:
- Fixed a bug where initializing a Quantity with a numpy array and a gpu backend would give CPUStorage
- raise TypeError if initializing a quantity with both a storage and a gt4py_backend argument
- eagerly create storage object when initializing Quantity
- make data type of quantity and storage reflect the gt4py_backend chosen, instead of being determined based on the data type being numpy/cupy

Other small fixes:
- use xarray's drop_vars instead of deprecated `drop` routine in `open_restart`
- add typing_extensions to requirements, was left out when added as import in previous PR

### CODE REVIEW

File: fv3gfs/util/_legacy_restart.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def prepend_label(filename, label=None):
 def load_partial_state_from_restart_file(
     file, restart_properties: RestartProperties, only_names=None
 ):
-    ds = xr.open_dataset(file).isel(Time=0).drop("Time")
+    ds = xr.open_dataset(file).isel(Time=0).drop_vars("Time")
```
Comment (mcgibbon) [inline fv3gfs/util/_legacy_restart.py:142]: Changed to avoid a DeprecationWarning from xarray.

File: requirements.txt [hunk 2]
```diff
@@ -31,3 +31,4 @@ toolz==0.10.0
 scipy==1.3.1
 git+https://github.com/GridTools/gt4py.git@db40777788863bfe87b9674b118e8bd5b8db0440#egg=gt4py
 cftime==1.2.1
+typing_extensions
```
Comment (mcgibbon) [inline requirements.txt:34]: Should have been added in an earlier PR where this was used.

File: fv3gfs/util/quantity.py [hunk 3]
```diff
@@ -269,12 +269,29 @@ def __init__(
             extent: number of points along each axis within the computational domain
             gt4py_backend: backend to use for gt4py storages, if not given this will
                 be derived from a Storage if given as the data argument, otherwise the
-                storage attribute is disabled and will raise an exception
+                storage attribute is disabled and will raise an exception. will raise
```
Comment (eddie-c-davis) [inline fv3gfs/util/quantity.py:272]: Minor point: should `exception. will` be `exception. Will`?

File: fv3gfs/util/quantity.py [hunk 4]
```diff
@@ -269,12 +269,29 @@ def __init__(
...
+                a TypeError if this is given with a gt4py storage type as data
         """
+        if origin is None:
+            origin = (0,) * len(dims)  # default origin at origin of array
+        else:
+            origin = tuple(origin)
+
+        if extent is None:
+            extent = tuple(length - start for length, start in zip(data.shape, origin))
```
Comment (eddie-c-davis) [inline fv3gfs/util/quantity.py:281]: Very nice extent computation!

File: fv3gfs/util/quantity.py [hunk 5]
```diff
@@ -371,41 +384,37 @@ def storage(self):
...
-            # must re-initialize compute domain view with new array
-            # this also can be removed when we merge GDP-3
-            self._compute_domain_view = BoundedArrayView(
-                self.data, self.dims, self.origin, self.extent
+                "gt4py backend was not specified when initializing this object"
             )
         return self._storage
 
+    def _initialize_storage(self, data, origin, gt4py_backend: str):
+        print(origin)
```
Comment (eddie-c-davis) [inline fv3gfs/util/quantity.py:394]: Rogue `print` statement?

File: fv3gfs/util/quantity.py [hunk 6]
```diff
@@ -371,41 +384,37 @@ def storage(self):
         this object, either by providing a Storage for data or explicitly specifying
         a backend.
         """
-        if gt4py is None:
-            raise ImportError("gt4py is not installed")
-        elif self._storage is None and self.gt4py_backend is None:
+        if self._storage is None:
```
Comment (eddie-c-davis) [inline fv3gfs/util/quantity.py:387]: Much cleaner!

**Review by eddie-c-davis — APPROVED** [review]
Couple of small points. This is a safer and simpler implementation overall. Nicely done.

File: fv3gfs/util/quantity.py (hunk 3 above)
Comment (mcgibbon) [reply fv3gfs/util/quantity.py:272]: Changed, I'm always unsure what to do when starting with lowercase and not using periods, but then running into a longer description that needs multiple sentences.

File: fv3gfs/util/quantity.py (hunk 5 above)
Comment (mcgibbon) [reply fv3gfs/util/quantity.py:394]: Good catch!

**Pull request merged**